libraries/lib-audio-io/AudioIO.cpp

```diff
@@ -2015,104 +2023,208 @@ bool AudioIO::ProcessPlaybackSlices(
          // warping
          if (frames > 0) {
             size_t produced = 0;
+            // Check for asynchronous user changes in mute, solo status
+            const auto discardable = SequenceShouldBeSilent(*mPlaybackSequences[iSequence]);
             if (toProduce)
                produced = mixer->Process(toProduce);
+
             //wxASSERT(produced <= toProduce);
             // Copy (non-interleaved) mixer outputs to one or more ring buffers
-            const auto nChannels = mPlaybackSequences[iSequence++]->NChannels();
+            const auto nChannels = mPlaybackSequences[iSequence]->NChannels();
+            
             for (size_t j = 0; j < nChannels; ++j) {
-               auto warpedSamples = mixer->GetBuffer(j);
-               const auto put = mPlaybackBuffers[iBuffer++]->Put(
-                  warpedSamples, floatSample, produced, frames - produced);
-               // wxASSERT(put == frames);
-               // but we can't assert in this thread
-               wxUnusedVar(put);
+
+               auto& buffer = mProcessingBuffers[iBuffer++];
+               const auto offset = buffer.size();
+               //there could be leftovers from the previous pass, don't discard them
+               buffer.resize(buffer.size() + frames, 0);
+
+               if(!discardable)
+               {
+                  auto warpedSamples = mixer->GetBuffer(j);
+                  std::copy_n(
+                     reinterpret_cast<const float*>(warpedSamples),
+                     produced,
+                     buffer.data() + offset);
+                  std::fill_n(
+                     buffer.data() + offset + produced,
+                     frames - produced,
+                     .0f);
+               }
+               else
+                  //TODO: mixer->Process() may be wasting CPU cycles
+                  //TODO: fade out smoothly
+                  std::fill_n(buffer.data() + offset, frames, 0);
```

Comment:
I notice that as part of this PR you have moved the logic to silence tracks (which might be driven by unpredictable presses of the mute buttons by the user) from the low-latency thread to this high-throughput thread.

Was that for simplification of the former, so that it does less work?  Or some other reason?

Does this comment mean you have not yet restored the microfading feature?  (I have not yet read enough to answer that for myself)

I think I like this change because as I recollect, the microfading had a dependency on the buffer size given to `audacityAudioCallback` which can depend on operating system or larency settings.  That should not be:  the duration of the fade in or out should be independent of that.

Reply:
> Was that for simplification of the former, so that it does less work? Or some other reason?

The reason is that we can't do master effects processing on low-latency thread and that processing should happen after all the tracks have been mixed into playback channels.